### ISSUE TITLE
reset() doesn't work for ExternalSource

### ISSUE THREAD

**Issue by Doom9234** [NVIDIA/DALI#2371]
`class ExternalSourcePipeline(Pipeline):

    def __init__(self, num_threads, device_id, gpu_prefetch_queue_depth, batch_size, prefetch_queue,img_token,,key_token,STOP_FLAG, part_list_queue, size_per_pipeline):
        super(ExternalSourcePipeline, self).__init__(
                                        num_threads=num_threads,
                                        device_id=device_id, 
                                        prefetch_queue_depth=gpu_prefetch_queue_depth,
                                        batch_size=batch_size
                                        )
        self.source_iter = self.get_queue_iter()
        # self.input = ops.ExternalSource(source=self.source_iter, num_outputs=2, cycle=True)
        self.input = ops.ExternalSource(source=self.get_queue_iter(), num_outputs=2, cycle=False)
        self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
        self.enhance = ops.BrightnessContrast(device="gpu", contrast=2)
        self.resize = ops.Resize(device="gpu", resize_x=224, resize_y=224)
        self.permute = ops.Transpose(device="gpu", output_layout=types.NCHW)
        self.cmnp = ops.CropMirrorNormalize(device="gpu",
                                            dtype=types.FLOAT,
                                            output_layout=types.NCHW,
                                            crop=(224, 224),
                                            mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                            std=[0.229 * 255,0.224 * 255,0.225 * 255])
        self.prefetch_queue = prefetch_queue
        self.img_token = img_token
        self.key_token = key_token
        self.STOP_FLAG= STOP_FLAG
        self.part_list_queue = part_list_queue
        self.iter_num = size_per_pipeline//self.batch_size

    def get_queue_iter(self):
        for _ in range(self.iter_num):
            raw_data = self.prefetch_queue.get() 
            print("current queue size is %s" % self.prefetch_queue.qsize())
            # images, keys = list(zip(*raw_data))
            data = [line.strip().split('\t') for line in raw_data]
            images = [np.frombuffer(base64.urlsafe_b64decode(d[self.img_token]), np.uint8) for d in data]
            keys = [np.array([d[self.key_token]], dtype=np.uint8) for d in data] 
            # images, keys = list(images), list(keys)
            yield (images, keys)

    def define_graph(self):
        images, keys = self.input()
        images = self.decode(images)
        output = self.enhance(images)
        output = self.resize(output)
        output = self.cmnp(output)
        return output, keys
    
    def reset(self):
        print("reseting")
        if self._last_iter:
            self._first_iter = True
            self._last_iter = False
            self._iter = 0
`

`train_loader = DALIGenericIterator(custom_pipeline, output_map=["data", "label"], size=-1, auto_reset=True)
`

when I tried to use train_loader after the first epoch, the reset() method seems not working..

**Comment by JanuszL** [comment]
Hi,
Indeed there is a bug in DALI, it should be fixed in #2373.
However, it won't allow you to reset the generator function you provide. What you can do instead is:
```
class ExternalSourcePipeline(Pipeline):

def __init__(self, num_threads, device_id, gpu_prefetch_queue_depth, batch_size, prefetch_queue,img_token,,key_token,STOP_FLAG, part_list_queue, size_per_pipeline):
    super(ExternalSourcePipeline, self).__init__(
                                    num_threads=num_threads,
                                    device_id=device_id, 
                                    prefetch_queue_depth=gpu_prefetch_queue_depth,
                                    batch_size=batch_size
                                    )
    self.source_iter = self.get_queue_iter()
    self.input = ops.ExternalSource(source=lambda: self.get_queue_iter(), num_outputs=2, cycle=False)
    self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
    self.enhance = ops.BrightnessContrast(device="gpu", contrast=2)
    self.resize = ops.Resize(device="gpu", resize_x=224, resize_y=224)
    self.permute = ops.Transpose(device="gpu", output_layout=types.NCHW)
    self.cmnp = ops.CropMirrorNormalize(device="gpu",
                                        dtype=types.FLOAT,
                                        output_layout=types.NCHW,
                                        crop=(224, 224),
                                        mean=[0.485 * 255,0.456 * 255,0.406 * 255],
                                        std=[0.229 * 255,0.224 * 255,0.225 * 255])
    self.prefetch_queue = prefetch_queue
    self.img_token = img_token
    self.key_token = key_token
    self.STOP_FLAG= STOP_FLAG
    self.part_list_queue = part_list_queue
    self.iter_num = size_per_pipeline//self.batch_size

def get_queue_iter(self):
     if self.i == self.iter_num:
         self.i = 0
         raise StopIteration
     self.i += 1
     raw_data = self.prefetch_queue.get() 
     print("current queue size is %s" % self.prefetch_queue.qsize())
     # images, keys = list(zip(*raw_data))
     data = [line.strip().split('\t') for line in raw_data]
     images = [np.frombuffer(base64.urlsafe_b64decode(d[self.img_token]), np.uint8) for d in data]
     keys = [np.array([d[self.key_token]], dtype=np.uint8) for d in data] 
     # images, keys = list(images), list(keys)
     return (images, keys)

def define_graph(self):
    images, keys = self.input()
    images = self.decode(images)
    output = self.enhance(images)
    output = self.resize(output)
    output = self.cmnp(output)
    return output, keys

def reset(self):
    print("reseting")
    if self._last_iter:
        self._first_iter = True
        self._last_iter = False
        self._iter = 0
```

**Comment by Doom9234** [comment]
@JanuszL Thank you for the quick reply, I think this close the issue.
